src/Juvix/Core/FromFrontend.hs

```diff
@@ -232,56 +276,125 @@ transformPatVar p = pure p
-transformTermHR (FE.List l) =
+transformTermHR _ (FE.List l) =
   throwFF $ ListUnimplemented l
-transformTermHR (FE.Tuple (FE.TupleLit es)) =
-  makeTuple <$> traverse transformTermHR es
-transformTermHR (FE.Block (FE.Bloc e)) =
-  transformTermHR e
-transformTermHR (FE.ExpRecord r) =
+transformTermHR q (FE.Tuple (FE.TupleLit es)) =
+  makeTuple <$> traverse (transformTermHR q) es
+transformTermHR q (FE.Block (FE.Bloc e)) =
+  transformTermHR q e
+transformTermHR _ (FE.ExpRecord r) =
   throwFF $ ExpRecordUnimplemented r
-transformTermHR (FE.ArrowE a) = transformArrow a
-transformTermHR (FE.NamedTypeE (FE.NamedType' _ e)) =
+transformTermHR q (FE.ArrowE a) = transformArrow q a
+transformTermHR q (FE.NamedTypeE (FE.NamedType' _ e)) =
   -- TODO the name will only become relevant (outside of arrows)
   -- when refinements are supported
-  transformTermHR e
-transformTermHR (FE.RefinedE r) =
+  transformTermHR q e
+transformTermHR _ (FE.RefinedE r) =
   throwFF $ RefinementsUnimplemented r
-transformTermHR (FE.UniverseName i) =
+transformTermHR _ (FE.UniverseName i) =
   -- TODO for universe polymorphism
   throwFF $ UniversesUnimplemented i
-transformTermHR (FE.Parened e) = transformTermHR e
-
-transformSimpleLet :: FE.Let -> Env primTy primVal (HR.Term primTy primVal)
-transformSimpleLet e@(FE.LetGroup name (clause :| []) body) = do
+transformTermHR q (FE.Parened e) = transformTermHR q e
+transformTermHR q (FE.DeclarationE (FE.DeclareExpression _ e)) =
+  transformTermHR q e
+
+transformApplication ::
+  NameSymbol.Mod ->
+  FE.Application ->
+  Env primTy primVal (HR.Term primTy primVal)
+transformApplication q (FE.App f xs) =
+  getSpecialE q f >>= flip go (toList xs)
+  where
+    go s xs = case s of
+      Just (ArrowS Nothing) -> do
+        ~[π, a, b] <- nargs 3 xs
+        π <- transformUsage q π
+        go (Just (ArrowS (Just π))) [a, b]
+      Just (ArrowS (Just π)) -> do
+        ~[xa, b] <- nargs 2 xs
+        (x, a) <- namedArg q xa
+        HR.Pi π x a <$> transformTermHR q b
+      Just (PairS Nothing) -> do
+        ~[π, a, b] <- nargs 3 xs
+        π <- transformUsage q π
+        go (Just (PairS (Just π))) [a, b]
+      Just (PairS (Just π)) -> do
+        ~[xa, b] <- nargs 2 xs
+        (x, a) <- namedArg q xa
+        HR.Sig π x a <$> transformTermHR q b
+      Just ColonS -> do
+        ~[a, b] <- nargs 2 xs
+        a <- transformTermHR q a
+        b <- transformTermHR q b
+        -- FIXME metavars for usage & universe
+        pure $ HR.Elim $ HR.Ann (Usage.SNat 1) a b 0
+      Just TypeS -> do
+        ~[i] <- nargs 1 xs
+        HR.Star <$> transformUniverse i
+      Just OmegaS ->
+        throwFF UnexpectedOmega
+      Nothing -> do
+        f' <- toElim f =<< transformTermHR q f
+        HR.Elim . foldl HR.App f' <$> traverse (transformTermHR q) xs
+    nargs n xs
+      | (xs, []) <- splitAt n (toList xs),
+        length xs == n =
+        pure xs
+      | otherwise =
+        throwFF $ WrongNumberBuiltinArgs n xs
```

Comment:
Application seems quite special

Reply:
Care to explain what all these special casings are?

Reply:
These are the interpretations of the magic builtins, e.g. `x : A`, which gets parsed as `App (:) [x, A]`, needs to be translated to `Ann x A` rather than ``(:) `App` x `App` A``.